### PR TITLE
HDDS-8462. Key in deleteTable uniqueness

### PR DESCRIPTION
**Pull request by sumitagrawl**
## What changes were proposed in this pull request?

1. delete key is appended with objectId to make it unique
2. Removed caching of deleteTable as its not used again due to uniqueness
and removed getting from deleteTable for old key (old key will not be present as delete key is unique)
3. This is to add repeated info and overwrite the same key again in deleteTable, which can cause missing cleanup of latest block appended as there KeyDeletingService has already run in parallel.

Uniqueness for multiple flows:

openKey cleanup: openKey appended with objectId
key commit(unused block and old version): openKey appended with objectId
delete key/keys: dbKey appended with objectId
directoryPurge: key appended with objectId

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8462

## How was this patch tested?

Unit test and integration test for delete behavior impact

### CODE REVIEW

**Comment by sumitagrawl** [comment]
Parent Issue: [HDDS-5905](https://issues.apache.org/jira/browse/HDDS-5905)
Related PR from which this split is created: https://github.com/apache/ozone/pull/4541
Old PR which is closed: https://github.com/apache/ozone/pull/3225

**Review by szetszwo — COMMENTED** [review]
@sumitagrawl , thanks a lot for splitting this from #4541 !

It turns out that object id is not unique for `OBJECT_ID_RECLAIM_BLOCKS`; see `OMKeyRequest.wrapUncommittedBlocksAsPseudoKey`.  We need to think about how to handle it.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMKeyCommitResponse.java
Comment (sumitagrawl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMKeyCommitResponse.java:124]: @szetszwo A tracking JIRA is already raised to handle this separately [HDDS-8467](https://issues.apache.org/jira/browse/HDDS-8467)

Comment (szetszwo) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMKeyCommitResponse.java:124]: @sumitagrawl , [HDDS-8467](https://issues.apache.org/jira/browse/HDDS-8467) is for `OBJECT_ID_RECLAIM_BLOCKS`.

But this comment is talking about `getOmKeyInfoList().get(0)`.  They are different problems.

Comment (szetszwo) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMKeyCommitResponse.java:124]: BTW, if this pull request has some known problem, we cannot merge it to the master branch even there are JIRAs tracking it.  We cannot introduce bugs to master branch and fix it later.

If there is a need, we may create a feature branch.  Then, we can merge some not-working intermediate code and then fix it before merging the feature branch to master.

Comment (sumitagrawl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMKeyCommitResponse.java:124]: @szetszwo 
In keyList, there will be possible 2 entry as keysToDelete.getOmKeyInfoList(),
- if VersionEnabled is false and old key is present, it will use old key objectId which will be unique.
- If above is not present for delete, and uncommited block is present, it will use OBJECT_ID_RECLAIM_BLOCKS which have value "0", can can have conflict key., This part will be handled as part of [HDDS-8467](https://issues.apache.org/jira/browse/HDDS-8467)

So this currently do not have issue and one issue will be handled as part of HDDS-8467

Comment (szetszwo) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMKeyCommitResponse.java:124]: First of all, I disagree that we can use the first key of the list (i,e, `getOmKeyInfoList().get(0)`) to represent the entire list.  Why not putting a deleteKey for each element in the list?

> ... one issue will be handled as part of [HDDS-8467](https://issues.apache.org/jira/browse/HDDS-8467)

As mentioned previously, we cannot introduce bugs to the master branch and fix it later.

Comment (aswinshakil) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMKeyCommitResponse.java:124]: Agree with @szetszwo. We might need to handle HDDS-8467, before committing this patch.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/AbstractOMKeyDeleteResponse.java
Comment (aswinshakil) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/AbstractOMKeyDeleteResponse.java:118]:
```suggestion
   * @param deleteKeyName (format: vol/buck/key/objectId)
```
Please also update `OmMetadataManagerImpl` with the schema change.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMKeyCommitResponse.java
Comment (sumitagrawl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMKeyCommitResponse.java]: @szetszwo @aswinshakil Updated... For old version, using obectId of old version, for uncommited block using trxLogIndex.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/AbstractOMKeyDeleteResponse.java
Comment (sumitagrawl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/AbstractOMKeyDeleteResponse.java]: Updated...

**Review by szetszwo — COMMENTED** [review]
@sumitagrawl , thanks for the update.  Please see the comments inlined.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java
Comment (sumitagrawl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java]: Updated...

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java
Comment (sumitagrawl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java]: Updated...

**Review by szetszwo — COMMENTED** [review]
The change looks good.  Just have a minor comment and a question inlined.

**Review by aswinshakil — COMMENTED** [review]
The changes mostly look good to me. I have a few questions, 

1. How does this handle upgrades? I guess older entries in the `deletedTable` before the upgrade will still be cleaned up by `KeyDeletingService`, while the newer entries after the upgrade use this new flow. If that's the case can we remove `getOldVersionsToCleanUp()` because there will be no older version with this new key format `/volume/bucket/key/objID`

2. Now with this change `RepeatedOmKeyInfo` will only have only entry right? As a part of Snapshot GC, we split the `RepeatedOmKeyInfo` into reclaimable keys and non-reclaimable keys. If it only has one entry we can avoid that step altogether.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequestWithFSO.java
Comment (sumitagrawl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequestWithFSO.java:222]: No, this is not required as, earlier, there is possibility for append entries as key name was same. Now key name is unique, so do not have this scenario.

**Comment by sumitagrawl** [comment]
> The changes mostly look good to me. I have a few questions,
> 
> 1. How does this handle upgrades? I guess older entries in the `deletedTable` before the upgrade will still be cleaned up by `KeyDeletingService`, while the newer entries after the upgrade use this new flow. If that's the case can we remove `getOldVersionsToCleanUp()` because there will be no older version with this new key format `/volume/bucket/key/objID`
> 2. Now with this change `RepeatedOmKeyInfo` will only have only entry right? As a part of Snapshot GC, we split the `RepeatedOmKeyInfo` into reclaimable keys and non-reclaimable keys. If it only has one entry we can avoid that step altogether.

@aswinshakil 
1. Related to upgrade, there is no impact to old entry already present in deletedTable as adding objectId is making it unique for any further addition of deleted key. Processing of deleted key will remain same for old version format and new format.

2. For Snapshot GC, we can not remove handling list as old version might be present having multiple entry. So still need to keep it same currently for compatibility.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyRequest.java
Comment (sumitagrawl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyRequest.java]: @szetszwo Updated...

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMTrashRecoverResponse.java
Comment (szetszwo) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMTrashRecoverResponse.java:61]: @sumitagrawl , It seems that `TrashRecover` won't work if there are old keys (without object id) in the deletedTable?

Comment (sumitagrawl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMTrashRecoverResponse.java:61]: @szetszwo This code is unused and needs to be removed, a JIRA [HDDS-8531](https://issues.apache.org/jira/browse/HDDS-8531) is created by @sadanand48 for handling same.

This does not have any problem.

Comment (szetszwo) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMTrashRecoverResponse.java:61]: If the code is unused, why changing it?

**Comment by szetszwo** [comment]
There is an error in https://github.com/apache/ozone/actions/runs/4867542164/jobs/8701924461?pr=4602#step:6:791 . Let's rerun the job.

**Review by szetszwo — APPROVED** [review]
@sumitagrawl , thanks a lot for working on this!

+1 the change looks good.

BTW, we should not keep using String as the db key since it is very easy to make mistakes.  We should add a `DeletedTableKey` class.

**Comment by szetszwo** [comment]
> Now with this change RepeatedOmKeyInfo will only have only entry right? ...

> .., we can not remove handling list as old version might be present having multiple entry. So still need to keep it same currently for compatibility.

@aswinshakil , @sumitagrawl , We may consider having a new table, say `DeletedTableWithObjectId`.   Then there is no compatibility issue.

**Pull request merged**
